lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -75,33 +105,113 @@ public String toString() {
+     * is set true.  Subsequent calls to <code>execute</code> will make a best-effort attempt
+     * to restart the queue.
+     * </nl>
+     */
     private static class ConcurrentExecutor implements CloseableExecutor {
-        private final Executor executor;
+        @NonNull
+        private final ThreadPoolExecutor executor;
+
+        @GuardedBy("this")
+        @NonNull
+        private final Queue<InstrumentedTask> pendingTasks = new LinkedList<>();
+
+        // a non-null stop latch is the flag that this executor has been stopped
+        @GuardedBy("this")
+        @Nullable
         private CountDownLatch stopLatch;
+
+        @GuardedBy("this")
         private int running;
 
-        private ConcurrentExecutor(Executor executor) { this.executor = executor; }
+        @GuardedBy("this")
+        private boolean needsRestart;
+
+        ConcurrentExecutor(@NonNull ThreadPoolExecutor executor) {
+            Preconditions.checkArgNotNull(executor, "executor");
+            this.executor = executor;
+        }
 
+        /**
+         * Schedule a task for concurrent execution.
+         * There are absolutely no guarantees about execution order, on this executor,
+         * particularly once it fails back to using the pending task queue.
+         * If there is insufficient room to schedule the task, safely, on the underlying
+         * executor, the task is added to the pendingTask queue and executed when space
+         * is available.
+         * This method may throw a <code>RejectedExecutionException</code> if the underlying
+         * executor's resources are completely exhausted even though <code>spaceAvailable</code>
+         * returns true.
+         *
+         * @param task a task for concurrent execution.
+         * @throws ExecutorClosedException    if the executor has been stopped
+         * @throws RejectedExecutionException if the underlying executor rejects the task
+         */
         @Override
-        public synchronized void execute(@NonNull Runnable task) {
-            if (stopLatch != null) { throw new ExecutorClosedExeception("Executor has been stopped"); }
+        public void execute(@NonNull Runnable task) {
+            Preconditions.checkArgNotNull(task, "task");
 
-            try {
-                executor.execute(new InstrumentedTask(task, this::finishTask));
-                running++;
-            }
-            catch (RejectedExecutionException e) {
-                dumpServiceState(executor, e, "size: " + running);
-                throw e;
+            final int pendingTaskCount;
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                if (spaceAvailable()) {
+                    if (needsRestart) { restartQueue(); }
+
+                    executeTask(new InstrumentedTask(task, this::finishTask));
```

Comment:
After restart the queue, do we need to check again if space is still available?

Reply:
It seems reasonable, to me, not to do that.  If, between the time that I check `spaceAvailable`, and the time I try to execute the task, the underlying executor is overwhelmed, then executeTask will throw.  That is the contract, whether `restartQueue` is called, in between, or not.

Reply:
I could think about one edge case that could cause the single pending task not running.
1. Assuming that each task submitted is a long running task. A lot of tasks get submitted until the space is unavailable.
2. One more task get submitted while the space is unavailable. The task get added to `pendingTasks`.
3. The restartQueue() is executed but because the space is unavailable, the reject exception is called, which will be ignored. Because there is no more task to come so the restartQueue() will not get called again so the single pending task will not be run.

Reply:
Beside that edge case, I think the implementation would work.

Reply:
I think that is not quite correct:  A task is never removed from the queue until it has been successfully submitted for execution, right?  I think that is always true.  If so, when the queue is restarted, by a subsequent call to `execute` the queue will be restarted and there will be another attempt to submit the head.